### ISSUE TITLE
Use reformatter.el to simplify ocamlformat.el

### ISSUE THREAD

**Issue by ludwigpacifici** [ocaml-ppx/ocamlformat#570]
Follow up from [issue 557](https://github.com/ocaml-ppx/ocamlformat/issues/557), https://github.com/ludwigpacifici/ocamlreformat.el is a proof of concept that aims to simplify the Elisp code of ocamlformat.el by using the library reformatter.el.

It preserves almost all the options of ocamlformat.el and should be a drop-in replacement.

`ocamlformat-show-errors` is left to do. It will probably require some discussions with the author of reformatter.el. I prefer to have some feedback first before raising an issue to handle the missing feature. Note reformatter.el provides some default error handling ; from the README:

> * Displaying error output
> * Colorising ANSI escape sequences in any error output

That would be great if the Emacs package is available on MELPA.

Let me know what are your thoughts on this proof of concept, or if I forgot other features from ocamlformat.el.

**Comment by jberdine** [comment]
Thanks! I am trying it locally for a while to see how it goes. It initially seems to work well here. I do already miss -show-errors though. :-)

It is very cool that formatting only a region "just works" so long as the region parses! Is there a function to bind to a key which automatically formats the region if it is active, and otherwise formats the buffer?

One minor nit from reading the source: if capitals are used, "OCaml" has been settled on, and we have been using "OCamlFormat". Also, I'm not a fan of having a non-nil lighter, but if there is one, should it have the current name that is not used elsewhere?

Could you comment on the use of insert-file-contents in reformatter.el versus replace-buffer-contents? Are we losing the benefits of replace-buffer-contents with this PR?

**Comment by ludwigpacifici** [comment]
Thank you for having a look at it :-)

> Is there a function to bind to a key which automatically formats the region if it is active, and otherwise formats the buffer?

I will update the README to show some basic configuration (probably with and without `use-package`). In the meantime the below should work:

```elisp
(add-to-list 'load-path "path_to_ocamlreformat")
(add-to-list 'load-path "path_to_reformatter")
(require 'ocamlreformat)
;; format the buffer on save
(add-hook 'tuareg-mode-hook #'ocamlreformat-on-save-mode)
;; bind TAB to format a valid region
(define-key tuareg-mode-map (kbd "<tab>") 'ocamlreformat-region)
```

> One minor nit from reading the source: if capitals are used, "OCaml" has been settled on, and we have been using "OCamlFormat"

Sure I'll do a pass to correct the naming.

> Also, I'm not a fan of having a non-nil lighter, but if there is one, should it have the current name that is not used elsewhere?

I used by default "OCamlFmt". It will show up like that:

![lighter](https://user-images.githubusercontent.com/923945/51940647-8455d000-240a-11e9-9456-b751ead455ce.png)

I don't believe this is used somewhere else (in the Emacs world). Lighter can be nil and will disappear (I am personally fine with that). However, the user cannot click on it to either turn off the minor mode or show the help. I can have a look at other implementations (Elm, Rust, JavaScript, etc.) and see there seems to have some consensus?

> Could you comment on the use of insert-file-contents in reformatter.el versus replace-buffer-contents?

I have not look at it yet. I'll reply to you later.

> Are we losing the benefits of replace-buffer-contents with this PR?

Same, I have not looked at the details. However, I noticed reformatter.el has no temporary buffer issue with Emacs 27.

ocamlreformat.el makes easy to add more argument support to be forwarded to OCamlFormat:

1. Define a new `defcustom`
2. Write a function that will convert the `defcustom` into string arguments for OCamlFormat
3. Call the function.

**Comment by jberdine** [comment]
Note that since ocamlformat cannot currently handle files that do not parse, it is still useful for people to have ocp-indent configured. So perhaps tab is not the best key binding. (I use `M-q` by analogy with filling in text modes.)

Re "OCamlFmt", it just seems strange that it is not the name of the mode, or the underlying tool, or anything else. My guess is that there would be less chance for confusion if a new name wasn't invented for this. And your argument about having a lighter basically being more discoverable is good.

**Comment by ludwigpacifici** [comment]
[`insert-file-contents`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Reading-from-Files.html) vs [`replace-buffer-contents`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Replacing.html): the latter seems to be preferred because it "attempts to keep point, markers, text properties, and overlays in the current buffer intact". However, the former is used with `replace` set to true, i.e. "it preserves some marker positions and it puts less data in the undo list". I am not an expert in this domain, but it feels roughly equivalent.

I added some minor changes to [ocamlreformat.el](https://github.com/ludwigpacifici/ocamlreformat.el) based on our discussion.

For information, I opened an [issue](https://github.com/purcell/reformatter.el/issues/3) with reformatter.el to be able to reproduce the behavior `ocamlformat-show-errors`. If this can be solved, would you be interested to have a PR that would "merge" the proof of concept into ocamlreformat.el? I guess this will raise questions about licensing (due to reformatter.el?) and having Elisp code dependency.

**Comment by jberdine** [comment]
I'm unsure about insert-file-contents vs replace-buffer-contents. It seems to be something that perhaps ought to be addressed at the reformatter.el level. But consider that the documentation of replace-buffer-contents explicitly lists the reformatting use case as one where replace-buffer-contents is preferable:
> One potential case where this behavior is useful is external code formatting programs

I guess that the licensing situation is unclear enough that it would be better to have GPL(-dependent) code in a separate repo. If we get to the point where an implementation based on reformatter.el can take over, I would not object to removing the emacs integration we have in the ocamlformat repo and pointing people to an external one. (Other maintainers might have other considerations though.)

**Comment by ludwigpacifici** [comment]
> I'm unsure about insert-file-contents vs replace-buffer-contents

Agreed, the documentation is not very clear. I asked the question to reformatter.el project (in the same opened issue).

**Comment by jberdine** [comment]
@hhugo, since you added the replace-buffer-contents support, would you or someone there care to comment on or try this PR in your setup? Do you see worse performance, etc. with this, which uses insert-file-contents instead? Note that this is still better than/different from the current ocamlformat.el behavior, for emacs <= 26.1, of manually applying the output of diff.

**Comment by ludwigpacifici** [comment]
I was curious about the performance, so I did some manual tests:

```
File             : ocamlformat/src/Fmt_ast.ml
Size             : 154,657 bytes
ocamlformat.el   : 0.62s (average with 5 runs)
ocamlreformat.el : 0.65s (average with 5 runs)

File             : ocamlformat/src/Ast.ml
Size             : 79,952 bytes
ocamlformat.el   : 0.24s (average with 5 runs)
ocamlreformat.el : 0.24s (average with 5 runs)

File             : ocamlformat/src/Conf.ml
Size             : 62,721 bytes
ocamlformat.el   : 0.13s (average with 5 runs)
ocamlreformat.el : 0.12s (average with 5 runs)

File             : ocaml/typing/typecore.ml
Size             : 199,975 bytes
ocamlformat.el   : 0.81s (average with 5 runs)
ocamlreformat.el : 0.69s (average with 5 runs)

File             : ocaml/parsing/lexer.ml
Size             : 171,425 bytes
ocamlformat.el   : 2.20s (average with 5 runs)
ocamlreformat.el : 0.11s (average with 5 runs)

File             : ocaml/typing/ctype.ml
Size             : 164,304 bytes
ocamlformat.el   : 0.81s (average with 5 runs)
ocamlreformat.el : 0.68s (average with 5 runs)

versions:
* GNU Emacs 27.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.7) of 2019-03-16
* ocamlformat 0.8
```

With my setup:
* ocamlformat.el used `replace-buffer-contents`
* ocamlreformat.el used `insert-file-contents`

How I measured performance:

1. Find large files. For example, [ocamlformat](https://github.com/ocaml-ppx/ocamlformat) or [ocaml](https://github.com/ocaml/ocaml), etc.: `find <project_path> -name '*.ml' | xargs du -b | sort -n | tail -n10`
2. Start an empty Emacs: `emacs -Q`
3. Load formatters and wrap them with a timer:

```elisp
(eval-when-compile
  (setq opam-share
        (substring
         (shell-command-to-string "opam config var share 2> /dev/null")
         0 -1)))
(add-to-list 'load-path (concat opam-share "/emacs/site-lisp"))
(require 'ocamlformat)

(add-to-list 'load-path "~/ocamlreformat.el/")
(add-to-list 'load-path "~/ocamlreformat.el/reformatter")
(require 'ocamlreformat)

(defmacro measure-time (&rest body)
  "Measure the time it takes to evaluate BODY."
  `(let ((time (current-time)))
     ,@body
     (message "%.06f" (float-time (time-since time)))))

(defun timed-ocamlformat ()
  (interactive)
  (measure-time (ocamlformat)))

(defun timed-ocamlreformat-buffer ()
  (interactive)
  (measure-time (ocamlreformat-buffer)))
```

4. Run `timed-ocamlreformat-buffer` and `timed-ocamlformat` for large files from step 1 and collect timings for statistics.

I am surprised to see that `ocamlreformat.el` is equivalent or faster than `ocamlformat.el`. I have not looked into details why there is such a gap for `ocaml/parsing/lexer.ml`. That would be interesting if someone can confirm or if I made a mistake.

**Comment by hhugo** [comment]
I expect the difference comes from insert-file-contents vs replace-buffer-contents.

Here is the documentation of replace-buffer-contents:
> As far as possible the replacement is non-destructive, i.e. existing
> buffer contents, markers, properties, and overlays in the current
> buffer stay intact.
> Warning: this function can be slow if there's a large number of small
> differences between the two buffers.

The performance alone is not really important. You also need to take into account the end-result.
In particular, replace-buffer-contents tries very hard to preserve the cursor and this makes the user experience very pleasant.

**Comment by ludwigpacifici** [comment]
FYI, author's point of view on `replace-buffer-contents`: https://github.com/purcell/reformatter.el/issues/10.

**Comment by gpetiot** [comment]
Can we get a TLDR for this issue? Are we waiting for something to be done in reformatter.el or do we just need to to some work in ocamlformat?

**Comment by ludwigpacifici** [comment]
TL;DR: The goal of this issue is to replace [ocamlformat.el](https://github.com/ocaml-ppx/ocamlformat/blob/master/emacs/ocamlformat.el) with the [reformatter.el](https://github.com/purcell/reformatter.el) library ; it would look like [this](https://github.com/ludwigpacifici/ocamlreformat.el/blob/master/ocamlreformat.el). I understand there is no work to be done on OCamlFormat. `reformatter.el` is usable as is but it has a slightly different behavior than `ocamlformat.el`.


Comparaison: pros of using `reformatter.el`

* See `reformatter.el` [rationale](https://github.com/purcell/reformatter.el#rationale)
* Usage of `insert-file-contents` instead of `replace-buffer-contents`
  * `replace-buffer-contents` is [unstable](https://bugzilla.redhat.com/show_bug.cgi?id=1597251) - However, I just noticed the bug has been solved with Emacs 26.2.
  * `replace-buffer-contents` [slighlty](https://github.com/ocaml-ppx/ocamlformat/issues/570#issuecomment-475992400) [slower](https://github.com/ocaml-ppx/ocamlformat/issues/570#issuecomment-476022692).

Cons:

* OCamlFormat will have a dependency on a third party library.
* `reformatter.el` does not have a proper way to mimic the behavior of [`ocamlformat-show-errors`](https://github.com/ocaml-ppx/ocamlformat/blob/master/emacs/ocamlformat.el#L51). It would rely on [prefix arguments](https://www.emacswiki.org/emacs/PrefixArgument). I am unsure how this [would work](https://github.com/purcell/reformatter.el/issues/3#issuecomment-486437156).

For information, I believe there is not much else I can do on that topic, so feel free to close the issue if you don't need the Elisp code replacement (if needed, it can be re-opened later).

**Comment by ludwigpacifici** [comment]
For information, `replace-buffer-contents` is integrated in [reformatter.el](https://github.com/purcell/reformatter.el/issues/10#issuecomment-493317764)

**Comment by jberdine** [comment]
As it has come up again recently, I want to note that there is an additional point in favor of this PR: it seems to (as in, seems to work but I have not tested much) support reformatting only the region, rather than only the entire buffer.

**Comment by ludwigpacifici** [comment]
Slowness with some pathological cases using `replace-buffer-contents` with reformatter.el: https://github.com/purcell/reformatter.el/issues/19

**Comment by purcell** [comment]
> In particular, replace-buffer-contents tries very hard to preserve the cursor and this makes the user experience very pleasant.

`insert-file-contents` with the `replace` flag does this too: the key difference is that `replace-buffer-contents` tries to preserve all markers, overlays etc. as well, which IMO is almost impossible to do in a generalised way, let alone doing it efficiently.

**Comment by gpetiot** [comment]
I finally took some time to try `ocamlreformat` with `reformatter`, it's nice but indeed as long as we can't see the errors in the buffer below and we have to manually select another buffer it won't replace the current plugin.
I think we can afford to remove the option we have in our plugin to only keep the output to the compilation buffer if that makes it simpler for reformatter.

**Comment by jberdine** [comment]
> I think we can afford to remove the option we have in our plugin to only keep the output to the compilation buffer

No, this is a necessary feature.

**Comment by purcell** [comment]
> I think we can afford to remove the option we have in our plugin to only keep the output to the compilation buffer if that makes it simpler for reformatter.

I've been thinking about how to accommodate this without overcomplicating `reformatter` or encouraging inconsistent behaviour of various `reformatter`-based libraries.

The `ocamlformat-buffer` function that `reformatter` would create already has a `display-buffer` prefix arg, which causes the error output buffer to be displayed. It defaults to  `nil`. If there were to be an `ocamlformat-display-errors` var or similar to influence  this, then you could globally set it to `t` to always display the errors buffer upon failure. Would that be sufficient? I know that ocamlformat can also display errors in the echo area, but I'd like to minimise variations.

**Comment by purcell** [comment]
> Do others know if there is some elisp config that would ensure that when reformatter wanted to display an error buffer, it would find where `*compilation*` is displayed, instead display the formatter's error buffer there, and then on the next successful formatter invocation, change back to displaying the `*compilation*` buffer there?

You could handle this with an appropriate entry in `display-buffer-alist`, I think. That's why generally I favour just calling `display-buffer`, since its behaviour is customisable. I would say that trying to control exactly where buffers get displayed is _really_ not something that a package like this should be doing, since that's very much a matter of personal preference (hence the existence of `display-buffer-alist`), and such a convenience for some users actively becomes a hindrance for others.

>  We have found the echo area to be too transient and small to be useful for this. I don't know if that functionality is used elsewhere, but it could go as far as I'm concerned.

Good to know, and I agree.

**Comment by gpetiot** [comment]
I'm marking this one as a "won't fix" and closing the issue.

If you think this will be a nice improvement to `ocamlformat.el` feel free to open a PR.
